servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Completable.java

```diff
@@ -757,8 +798,48 @@ public final Completable beforeCancel(Runnable onCancel) {
      * </ul>
      * for Subscriptions/{@link Subscriber}s of the returned {@link Completable}. <strong>MUST NOT</strong> throw.
      * @return The new {@link Completable}.
+     *
+     * @see <a href="http://reactivex.io/documentation/operators/do.html">ReactiveX do operator.</a>
      */
     public final Completable beforeFinally(Runnable doFinally) {
+        return beforeFinally(new RunnableTerminalSignalConsumer(doFinally));
+    }
+
+    /**
+     * Invokes the corresponding method on {@code beforeFinally} {@link TerminalSignalConsumer} argument
+     * <strong>before</strong> any of the following terminal methods are called:
+     * <ul>
+     *     <li>{@link Subscriber#onComplete()} - invokes {@link TerminalSignalConsumer#onComplete()}</li>
+     *     <li>{@link Subscriber#onError(Throwable)} - invokes {@link TerminalSignalConsumer#onError(Throwable)}</li>
+     *     <li>{@link Cancellable#cancel()} - invokes {@link TerminalSignalConsumer#cancel()}</li>
+     * </ul>
+     * for Subscriptions/{@link Subscriber}s of the returned {@link Completable}.
+     * <p>
+     * From a sequential programming point of view this method is roughly equivalent to the following:
+     * <pre>{@code
+     *  try {
+     *      resultOfThisCompletable();
+     *  } catch(Throwable t) {
+     *      doFinally.onError(t);
```

Comment:
I used this pattern for `beforeFinally(TerminalSignalConsumer)` for Completable, Single, and Publisher as it represents logic of `BeforeFinallyCompletableSubscriber`. 
`whenFinally(TerminalSignalConsumer)` and `afterFinally(TerminalSignalConsumer)` use another control flow, like in `AfterFinallyCompletableSubscriber`:
```java
@Override
public void onComplete() {
    try {
        original.onComplete();
    } finally {
        doFinally.onComplete();
    }
}
```
Javadoc for `Completable.afterFinally(TerminalSignalConsumer)` written in the same way:
```
     * <pre>{@code
     *  try {
     *      resultOfThisCompletable();
     *  } finally {
     *      nextOperation(); // Maybe notifying of cancellation, or termination
     *      doFinally.onTerminalSignal(); // Invokes the corresponding terminal signal
     *  }
     * }</pre>
```
We do not try-catch for `afterFinally`/`whenFinally` and always touch `TerminalSignalConsumer` in `finally` block.
I use `doFinally.onTerminalSignal()` instead of `doFinally.onCompete()` because the same patters works for `onError` and `cancel`.

Can you please clarify how the javadoc for `afterFinally`/`whenFinally` should change?

Reply:
Discussed offline, the control flow in javadoc is not related to the actual implementation of `Subscriber` and should represent a similar behavior for sequential programming. Updated.